File: 48-rotate-image/rotate-image.java
class Solution {
    static void reverse(int a[], int l, int r){
        if(l>r){
            return;
        }

        int temp = a[l];
        a[l] = a[r];
        a[r] = temp;

        reverse(a, l+1, r-1);
    }
    public void rotate(int[][] matrix) {
        for(int i=0; i<matrix.length; i++){
            for(int j=0; j<matrix[0].length; j++){
                if(i<=j){
                    int temp = matrix[i][j];
                    matrix[i][j] = matrix[j][i];
                    matrix[j][i]= temp;
                }
            }
        }

        for(int i=0; i<matrix.length; i++){
            reverse(matrix[i], 0, matrix.length-1);
        }

    }
}

File: 48-rotate-image/rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        for(int i=0;i<rows;i++){
            for(int j=i;j<cols;j++){
                swap(matrix[i][j],matrix[j][i]);         
            }
        }
        for(int i=0;i<rows;i++){
            for(int j=0;j<cols/2;j++){
                swap(matrix[i][j],matrix[i][cols-1-j]);
            }            
        }

    }
};